common/src/main/java/io/seata/common/loader/EnhancedServiceLoader.java

```diff
@@ -198,181 +173,446 @@
      */
     @SuppressWarnings("rawtypes")
     static <S> List<Class> getAllExtensionClass(Class<S> service, ClassLoader loader) {
-        return findAllExtensionClass(service, null, loader);
+        return InnerEnhancedServiceLoader.getServiceLoader(service).getAllExtensionClass(loader);
     }
 
-    private static <S> S loadFile(Class<S> service, String activateName, ClassLoader loader) {
-        return loadFile(service, activateName, loader, null, null);
-    }
 
-    @SuppressWarnings("rawtypes")
-    private static <S> S loadFile(Class<S> service, String activateName, ClassLoader loader, Class[] argTypes,
-                                  Object[] args) {
-        try {
-            boolean foundFromCache = true;
-            List<Class> extensions = providers.get(service);
-            if (extensions == null) {
-                synchronized (service) {
-                    extensions = providers.get(service);
-                    if (extensions == null) {
-                        extensions = findAllExtensionClass(service, activateName, loader);
-                        foundFromCache = false;
-                        providers.put(service, extensions);
-                    }
+    private static class InnerEnhancedServiceLoader<S> {
+        private static final Logger LOGGER = LoggerFactory.getLogger(EnhancedServiceLoader.class);
+        private static final String SERVICES_DIRECTORY = "META-INF/services/";
+        private static final String SEATA_DIRECTORY = "META-INF/seata/";
+
+        private static final ConcurrentMap<Class<?>, InnerEnhancedServiceLoader<?>> SERVICE_LOADERS =
+                new ConcurrentHashMap<>();
+
+        private final Class<S> type;
+        private final Holder<List<ExtensionDefinition>> definitionsHolder = new Holder<>();
+        private final ConcurrentMap<ExtensionDefinition, Holder<Object>> definitionToInstanceMap =
+                new ConcurrentHashMap<>();
+        private final ConcurrentMap<String, List<ExtensionDefinition>> nameToDefinitionsMap = new ConcurrentHashMap<>();
```

Comment:
When is this a collection instead of a single element?

Reply:
I think we can allow an extension implementation to have the same name and interface name, so here a name may correspond to multiple implementations